File: core/workers/musicbrainz.py
"""
musicbrainz.py — MusicBrainz enrichment worker.

Gebruikt de musicbrainzngs bibliotheek (ingebouwde rate limit: 1 req/sec).
Geen API key nodig. User-Agent verplicht.
"""

import logging
import time
from typing import Any, Optional

import musicbrainzngs

from core.workers.base import BaseWorker

musicbrainzngs.set_useragent("LastfmMuziekApp", "2.0", "muziek-dashboard")
musicbrainzngs.set_rate_limit(limit_or_interval=1.0, new_requests=1)


class MusicBrainzWorker(BaseWorker):
    source = "musicbrainz"

    def process(self, item: dict) -> dict[str, Any]:
        try:
            entity_type = item["entity_type"]
            name        = item["entity_name"]

            if entity_type == "artist":
                data = self._process_artist(name)
            elif entity_type == "album":
                data = self._process_album(name)
            elif entity_type == "track":
                data = self._process_track(name)
            else:
                return {"ok": False, "error": f"Onbekend entity_type: {entity_type}"}

            if not data:
                return {"ok": False, "error": "Geen MBZ-resultaat gevonden"}
            return {"ok": True, "data": data}

        except Exception as exc:
            self.log.warning(f"MusicBrainz worker mislukt voor '{item.get('entity_name')}': {exc}")
            return {"ok": False, "error": str(exc)}

    # ── Artiest ───────────────────────────────────────────────────────────────

    def _process_artist(self, name: str) -> Optional[dict]:
        result = musicbrainzngs.search_artists(artist=name, limit=5)
        artists = result.get("artist-list", [])
        if not artists:
            return None

        best = self.fuzzy_best(artists, name, lambda a: a.get("name", ""))
        if not best:
            best = artists[0]

        return {
            "mbid":           best.get("id"),
            "entity_type":    "artist",
            "name":           best.get("name"),
            "sort_name":      best.get("sort-name"),
            "disambiguation": best.get("disambiguation"),
            "country":        best.get("country"),
            "area":           (best.get("area") or {}).get("name"),
            "type":           best.get("type"),
            "gender":         best.get("gender"),
            "begin_date":     (best.get("life-span") or {}).get("begin"),
            "end_date":       (best.get("life-span") or {}).get("end"),
            "ended":          (best.get("life-span") or {}).get("ended") == "true",
            "tags":           [t["name"] for t in sorted(
                                   best.get("tag-list", []),
                                   key=lambda t: -int(t.get("count", 0))
                               )][:10],
            "isni_list":      best.get("isni-list", [])[:3],
            "source":         "musicbrainz",
            "fetchedAt":      int(time.time() * 1000),
        }

    # ── Album (release-group) ─────────────────────────────────────────────────

    def _process_album(self, name: str) -> Optional[dict]:
        result = musicbrainzngs.search_release_groups(releasegroup=name, limit=5)
        groups = result.get("release-group-list", [])
        if not groups:
            return None

        best = self.fuzzy_best(groups, name, lambda g: g.get("title", ""))
        if not best:
            best = groups[0]

        return {
            "mbid":               best.get("id"),
            "entity_type":        "release",
            "title":              best.get("title"),
            "disambiguation":     best.get("disambiguation"),
            "primary_type":       best.get("primary-type"),
            "secondary_types":    best.get("secondary-type-list", []),
            "first_release_date": best.get("first-release-date"),
            "artist_credit":      [
                (ac.get("artist") or {}).get("name")
                for ac in best.get("artist-credit", [])
                if isinstance(ac, dict)
            ],
            "tags":               [t["name"] for t in sorted(
                                       best.get("tag-list", []),
                                       key=lambda t: -int(t.get("count", 0))
                                   )][:8],
            "source":             "musicbrainz",
            "fetchedAt":          int(time.time() * 1000),
        }

    # ── Track (recording) ─────────────────────────────────────────────────────

    def _process_track(self, name: str) -> Optional[dict]:
        result = musicbrainzngs.search_recordings(recording=name, limit=5)
        recordings = result.get("recording-list", [])
        if not recordings:
            return None

        best = self.fuzzy_best(recordings, name, lambda r: r.get("title", ""))
        if not best:
            best = recordings[0]

        return {
            "mbid":           best.get("id"),
            "entity_type":    "recording",
            "title":          best.get("title"),
            "disambiguation": best.get("disambiguation"),
            "length":         best.get("length"),
            "artist_credit":  [
                (ac.get("artist") or {}).get("name")
                for ac in best.get("artist-credit", [])
                if isinstance(ac, dict)
            ],
            "isrcs":          best.get("isrc-list", [])[:3],
            "tags":           [t["name"] for t in sorted(
                                   best.get("tag-list", []),
                                   key=lambda t: -int(t.get("count", 0))
                               )][:8],
            "releases":       [
                {"mbid": r.get("id"), "title": r.get("title"), "date": r.get("date")}
                for r in best.get("release-list", [])[:3]
            ],
            "source":         "musicbrainz",
            "fetchedAt":      int(time.time() * 1000),
        }

File: .claude/worktrees/distracted-saha-2b56b7/services/enrichment/workers/musicbrainz.js
'use strict';
// ── MusicBrainz enrichment worker ─────────────────────────────────────────────
// API: https://musicbrainz.org/ws/2  — geen API key nodig
// Rate limit: strikt 1 req/sec (MusicBrainz-beleid)
// User-Agent: verplicht in formaat 'AppName/Version (contact@email)'
// Hergebruikt de rate-limit queue uit services/musicbrainz.js via mbzGet

const { mbzGet } = require('../../musicbrainz');

const MBZ_APP_UA = 'LastfmMuziekApp/2.0 (muziek-dashboard)';

// ── Fuzzy helper ──────────────────────────────────────────────────────────────
function _fuzzyScore(a, b) {
  a = (a || '').toLowerCase().trim();
  b = (b || '').toLowerCase().trim();
  if (a === b) return 1.0;
  if (a.length < 2 || b.length < 2) return 0.0;
  const bigrams = (s) => {
    const m = new Map();
    for (let i = 0; i < s.length - 1; i++) {
      const bg = s.slice(i, i + 2);
      m.set(bg, (m.get(bg) || 0) + 1);
    }
    return m;
  };
  const aMap = bigrams(a);
  const bMap = bigrams(b);
  let intersection = 0;
  for (const [bg, cnt] of aMap) {
    intersection += Math.min(cnt, bMap.get(bg) || 0);
  }
  const total = a.length + b.length - 2;
  return total > 0 ? (2 * intersection) / total : 0;
}

class MusicBrainzWorker {
  constructor(db, log) {
    this.db  = db;
    this.log = log.child ? log.child({ worker: 'musicbrainz' }) : log;
  }

  // ── Hoofdverwerker ────────────────────────────────────────────────────────

  async process(entity) {
    try {
      const type = entity.entity_type;
      let data;

      if (type === 'artist') {
        data = await this._processArtist(entity.entity_name);
      } else if (type === 'album') {
        data = await this._processAlbum(entity.entity_name);
      } else if (type === 'track') {
        data = await this._processTrack(entity.entity_name);
      } else {
        return { ok: false, error: `Onbekend entity_type: ${type}` };
      }

      if (!data) return { ok: false, error: 'Geen MBZ-resultaat gevonden' };
      return { ok: true, data };
    } catch (err) {
      this.log.warn({ err: err.message, entity: entity.entity_name }, 'MusicBrainz worker mislukt');
      return { ok: false, error: err.message };
    }
  }

  // ── Artiest ───────────────────────────────────────────────────────────────

  async _processArtist(name) {
    const q    = encodeURIComponent(`artist:"${name.replace(/"/g, '')}"`);;
    const data = await mbzGet(`/artist?query=${q}&limit=5&fmt=json`);
    const list = data.artists || [];
    if (!list.length) return null;

    const norm  = s => (s || '').toLowerCase().trim();
    const exact = list.find(a => norm(a.name) === norm(name));
    let best    = exact;
    if (!best) {
      best = list.reduce((b, a) => {
        const score = _fuzzyScore(a.name, name);
        return score > _fuzzyScore(b.name, name) ? a : b;
      }, list[0]);
    }
    if (!best) return null;

    return {
      mbid:           best.id,
      entity_type:    'artist',
      name:           best.name,
      sort_name:      best['sort-name'] || null,
      disambiguation: best.disambiguation || null,
      country:        best.country || null,
      area:           best.area?.name || null,
      type:           best.type || null,
      gender:         best.gender || null,
      begin_date:     best['life-span']?.begin || null,
      end_date:       best['life-span']?.end || null,
      ended:          best['life-span']?.ended || false,
      tags:           (best.tags || [])
                        .sort((a, b) => b.count - a.count)
                        .slice(0, 10)
                        .map(t => t.name),
      isni_list:      (best.isnis || []).slice(0, 3),
      source:         'musicbrainz',
      fetchedAt:      Date.now(),
    };
  }

  // ── Album (release-group) ─────────────────────────────────────────────────

  async _processAlbum(name) {
    const q    = encodeURIComponent(`releasegroup:"${name.replace(/"/g, '')}"`);
    const data = await mbzGet(`/release-group?query=${q}&limit=5&fmt=json`);
    const list = data['release-groups'] || [];
    if (!list.length) return null;

    const norm  = s => (s || '').toLowerCase().trim();
    const exact = list.find(rg => norm(rg.title) === norm(name));
    const best  = exact || list[0];
    if (!best) return null;

    return {
      mbid:           best.id,
      entity_type:    'release',
      title:          best.title,
      disambiguation: best.disambiguation || null,
      primary_type:   best['primary-type'] || null,
      secondary_types: (best['secondary-types'] || []),
      first_release_date: best['first-release-date'] || null,
      artist_credit:  (best['artist-credit'] || []).map(ac => ac.artist?.name).filter(Boolean),
      tags:           (best.tags || [])
                        .sort((a, b) => b.count - a.count)
                        .slice(0, 8)
                        .map(t => t.name),
      source:         'musicbrainz',
      fetchedAt:      Date.now(),
    };
  }

  // ── Track (recording) ─────────────────────────────────────────────────────

  async _processTrack(name) {
    const q    = encodeURIComponent(`recording:"${name.replace(/"/g, '')}"`);
    const data = await mbzGet(`/recording?query=${q}&limit=5&fmt=json`);
    const list = data.recordings || [];
    if (!list.length) return null;

    const norm  = s => (s || '').toLowerCase().trim();
    const exact = list.find(r => norm(r.title) === norm(name));
    const best  = exact || list[0];
    if (!best) return null;

    return {
      mbid:           best.id,
      entity_type:    'recording',
      title:          best.title,
      disambiguation: best.disambiguation || null,
      length:         best.length ?? null,
      artist_credit:  (best['artist-credit'] || []).map(ac => ac.artist?.name).filter(Boolean),
      isrcs:          (best.isrcs || []).slice(0, 3),
      tags:           (best.tags || [])
                        .sort((a, b) => b.count - a.count)
                        .slice(0, 8)
                        .map(t => t.name),
      releases:       (best.releases || []).slice(0, 3).map(r => ({
        mbid:  r.id,
        title: r.title,
        date:  r.date || null,
      })),
      source:         'musicbrainz',
      fetchedAt:      Date.now(),
    };
  }
}

module.exports = { MusicBrainzWorker };
